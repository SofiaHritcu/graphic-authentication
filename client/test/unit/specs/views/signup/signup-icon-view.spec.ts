import { shallowMount } from "@vue/test-utils";
import "regenerator-runtime/runtime";
import SignUpIconView from "@/views/signup/signup-icon-view.vue";
import mockStore from "../../store/mocked-store-modules";
import {
  iconCategoriesSortedMock,
  iconsCategoriesMock,
} from "../../store/modules/icons/mock-data/mock";

describe("Login Icon View", () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallowMount(SignUpIconView, {
      store: mockStore,
      data: {
        userNameRules: [],
        iconPassCount: 6,
        iconPasses: [],
        iconPassLastCompletedIndex: 0,
        loginFormRef: "loginFormRef",
      },
      computed: {
        onLargerViewPort: () => true,
      },
      mocks: {
        $router: {
          push: jest.fn(),
        },
        $vuetify: {
          breakpoint: {
            name: "",
          },
        },
      },
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  describe("snapshot", () => {
    it("should match snapshot", () => {
      expect(wrapper.vm.$el).toMatchSnapshot();
    });
  });

  describe("data rules", () => {
    describe("nameRules", () => {
      it("should return a message if the name is not defined", () => {
        var nameRules = wrapper.vm.nameRules;
        expect(nameRules[0]("")).toBe("name is required");
      });
      it("should return a message if the name is longer than 20 characters", () => {
        var nameRules = wrapper.vm.nameRules;
        expect(nameRules[1]("abcdefghijklmnopqrstuvwxyz")).toBe(
          "the name must be less than 20 characters"
        );
      });
    });
    describe("userNameRules", () => {
      it("should return a message if the userName is not defined", () => {
        var userNameRules = wrapper.vm.userNameRules;
        expect(userNameRules[0]("")).toBe("username is required");
      });
      it("should return a message if the userName is longer than 10 characters", () => {
        var userNameRules = wrapper.vm.userNameRules;
        expect(userNameRules[1]("abcdefghijklm")).toBe(
          "username must be less than 10 characters"
        );
      });
    });
    describe("emailRules", () => {
      it("should return a message if the email is not defined", () => {
        var emailRules = wrapper.vm.emailRules;
        expect(emailRules[0]("")).toBe("e-mail is required");
      });
      it("should return a message if the email is not in a correct format", () => {
        var emailRules = wrapper.vm.emailRules;
        expect(emailRules[1]("abcdefghijklm")).toBe("e-mail must be valid");
      });
    });
  });

  describe("methods", () => {
    describe("handleBackBtnClick", () => {
      it("should call router push with the corresponding parameters", async () => {
        expect.assertions(1);
        const spyRouterPush = jest.spyOn(wrapper.vm.$router, "push");

        await wrapper.vm.handleBackBtnClick();

        expect(spyRouterPush).toHaveBeenCalledWith({ name: "ga-signup" });
      });
    });
    describe("handleIconCategoryChange", () => {
      it("should make the logo invisible and set the current icons to the corresponding value", async () => {
        expect.assertions(2);

        await wrapper.setData({
          logoVisible: true,
        });

        wrapper.vm.handleIconCategoryChange(iconsCategoriesMock[0].category);

        expect(wrapper.vm.logoVisible).toBe(false);
        expect(wrapper.vm.iconItems).toEqual(iconsCategoriesMock[0].icons);
      });
    });
    describe("handleIconClick", () => {
      it("should make the logo invisible and update the icon passes if the pass is not long enough", async () => {
        expect.assertions(2);

        await wrapper.setData({
          logoVisible: true,
        });
        const iconItem = iconsCategoriesMock[0].icons[0];

        wrapper.vm.handleIconClick(iconItem);

        expect(wrapper.vm.iconPasses).toEqual({
          "0": iconItem,
          "1": "",
          "2": "",
          "3": "",
          "4": "",
          "5": "",
        });
        expect(wrapper.vm.iconPassLastCompletedIndex).toBe(1);
      });
    });
    it("should check the password strength if the pass is complete", async () => {
      expect.assertions(3);

      await wrapper.setData({
        logoVisible: true,
        iconPassLastCompletedIndex: 5,
        iconPasses: {
          "0": "a",
          "1": "a",
          "2": "a",
          "3": "a",
          "4": "a",
          "5": "",
        },
      });
      const iconItem = iconCategoriesSortedMock[0].icons[0];

      wrapper.vm.handleIconClick(iconItem);

      expect(wrapper.vm.iconPasses).toEqual({
        "0": "a",
        "1": "a",
        "2": "a",
        "3": "a",
        "4": "a",
        "5": iconItem,
      });
      expect(wrapper.vm.iconPassLastCompletedIndex).toBe(6);
      expect(wrapper.vm.iconPassStrengthInterval).toStrictEqual({
        lower: 0,
        upper: 100,
        qualifying: "bad level of security",
        color: "#C62828",
        vColor: "red",
        vOpacity: "darken-3",
      });
    });
    describe("handleClearBtnClick", () => {
      it("should reset all data props related to the pass", () => {
        const spySetUpDefaultIconsPass = jest.spyOn(
          wrapper.vm,
          "setUpDefaultIconsPass"
        );
        wrapper.vm.handleClearBtnClick();

        expect(wrapper.vm.userName).toBe("");
        expect(wrapper.vm.iconPasses).toStrictEqual(["", "", "", "", "", ""]);
        expect(wrapper.vm.iconPassLastCompletedIndex).toBe(0);
        expect(spySetUpDefaultIconsPass).toHaveBeenCalled();
      });
    });
    describe("submitSignup", () => {
      it("should call fetchSignup with the user input", async () => {
        expect.assertions(1);

        const spyOnFetchSignup = jest.spyOn(wrapper.vm, "fetchSignup");
        await wrapper.setData({
          name: "nameTest",
          userName: "userNameTest",
          email: "email@test.com",
          passAfterTransformation: "passwordTest",
        });

        await wrapper.vm.submitSignup();

        expect(spyOnFetchSignup).toHaveBeenCalledWith({
          name: "nameTest",
          userName: "userNameTest",
          email: "email@test.com",
          password: "passwordTest",
          confirmPassword: "passwordTest",
        });
      });
    });
    describe("handleSignupBtnClick", () => {
      it("should submit signup if the form is valid", () => {
        wrapper.vm.$refs.signupFormRef.validate = jest.fn(() => true);
        const spySubmitSignup = jest.spyOn(wrapper.vm, "submitSignup");

        wrapper.vm.handleSignupBtnClick();

        expect(spySubmitSignup).toHaveBeenCalled();
      });
      it("should reset the form if it is not valid", () => {
        wrapper.vm.$refs.signupFormRef.validate = jest.fn(() => false);
        wrapper.vm.$refs.signupFormRef.reset = jest.fn();
        const spyFormReset = jest.spyOn(
          wrapper.vm.$refs.signupFormRef,
          "reset"
        );

        wrapper.vm.handleSignupBtnClick();

        expect(wrapper.vm.signupErrorShows).toBe(true);
        expect(spyFormReset).toHaveBeenCalled();
      });
    });
  });

  //   describe("computed", () => {
  //     describe("iconsScrollSize", () => {
  //       it("should return 400 when being on larger viewports", () => {
  //         wrapper.vm.$vuetify.breakpoint.name = "lg";
  //         expect(wrapper.vm.iconsScrollSize).toEqual(400);
  //       });
  //       it("should return 200 when being on smaller viewports", () => {
  //         wrapper.vm.$vuetify.breakpoint.name = "lg";
  //         expect(wrapper.vm.iconsScrollSize).toEqual(400);
  //       });
  //     });
  //     describe("onLargerViewPort", () => {
  //       it("should return true for lg or xl breakpoint", () => {
  //         wrapper.vm.$vuetify.breakpoint.name = "lg";
  //         expect(wrapper.vm.onLargerViewPort).toEqual(true);
  //       });
  //     });
});
