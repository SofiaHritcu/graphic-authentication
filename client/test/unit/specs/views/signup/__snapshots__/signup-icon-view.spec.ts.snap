// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Icon View snapshot should match snapshot 1`] = `
<div
  class="ga__signup_icon_view"
>
  <v-row
    class="ga__signup_icon_view__title_row"
  >
    <v-btn
      color="teal darken-2"
      outlined=""
    >
      
        Go Back To Choose IPASS
        
      <fa-icon
        class="ml-2"
        color="#00796B"
        icon="fa-solid fa-arrow-rotate-left"
        size="lg"
      />
    </v-btn>
  </v-row>
   
  <v-row
    align-content="center"
    class="ga__signup_icon_view__title"
    justify="center"
  >
    <v-col
      cols="12"
    >
      <div
        class="text-overline white--text text--darken-1 ga__signup_icon_view__title__text"
      >
        
          Signup
        
      </div>
    </v-col>
  </v-row>
   
  <v-row
    class="ga__signup_icon_view__content"
  >
    <v-col
      cols="12"
      lg="6"
    >
      <v-form>
        <v-row
          class="ga__signup_icon_view__form_row"
          justify="center"
        >
          <v-col
            class="ga__signup_icon_view__form_col"
            cols="8"
          >
            <v-text-field
              class="ga__signup_icon_view__form_field"
              color="teal"
              counter="20"
              dark=""
              filled=""
              label="first and last name"
              required=""
              rules="(v) => {
          /* istanbul ignore next */
          cov_203uahwopn().f[2]++;
          cov_203uahwopn().s[12]++;
          return (
            /* istanbul ignore next */
            (cov_203uahwopn().b[0][0]++, !!v) ||
            /* istanbul ignore next */
            (cov_203uahwopn().b[0][1]++, \\"name is required\\")
          );
        },(v) => {
          /* istanbul ignore next */
          cov_203uahwopn().f[3]++;
          cov_203uahwopn().s[13]++;
          return (
            /* istanbul ignore next */
            (cov_203uahwopn().b[1][0]++, v.length <= 20) ||
            /* istanbul ignore next */
            (cov_203uahwopn().b[1][1]++, \\"the name must be less than 20 characters\\")
          );
        }"
            />
          </v-col>
        </v-row>
         
        <v-row
          class="ga__signup_icon_view__form_row"
          justify="center"
        >
          <v-col
            class="ga__signup_icon_view__form_col"
            cols="8"
          >
            <v-text-field
              class="ga__signup_icon_view__form_field"
              color="teal"
              counter="10"
              dark=""
              filled=""
              label="user name"
              required=""
              rules="(v) => {
          /* istanbul ignore next */
          cov_203uahwopn().f[4]++;
          cov_203uahwopn().s[14]++;
          return (
            /* istanbul ignore next */
            (cov_203uahwopn().b[2][0]++, !!v) ||
            /* istanbul ignore next */
            (cov_203uahwopn().b[2][1]++, \\"username is required\\")
          );
        },(v) => {
          /* istanbul ignore next */
          cov_203uahwopn().f[5]++;
          cov_203uahwopn().s[15]++;
          return (
            /* istanbul ignore next */
            (cov_203uahwopn().b[3][0]++, v.length <= 10) ||
            /* istanbul ignore next */
            (cov_203uahwopn().b[3][1]++, \\"username must be less than 10 characters\\")
          );
        }"
            />
          </v-col>
        </v-row>
         
        <v-row
          class="ga__signup_icon_view__form_row"
          justify="center"
        >
          <v-col
            class="ga__signup_icon_view__form_col"
            cols="8"
          >
            <v-text-field
              class="ga__signup_icon_view__form_field"
              color="teal"
              dark=""
              filled=""
              label="email"
              required=""
              rules="(v) => {
          /* istanbul ignore next */
          cov_203uahwopn().f[6]++;
          cov_203uahwopn().s[16]++;
          return (
            /* istanbul ignore next */
            (cov_203uahwopn().b[4][0]++, !!v) ||
            /* istanbul ignore next */
            (cov_203uahwopn().b[4][1]++, \\"e-mail is required\\")
          );
        },(v) => {
          /* istanbul ignore next */
          cov_203uahwopn().f[7]++;
          cov_203uahwopn().s[17]++;
          return (
            /* istanbul ignore next */
            (cov_203uahwopn().b[5][0]++, /.+@.+/.test(v)) ||
            /* istanbul ignore next */
            (cov_203uahwopn().b[5][1]++, \\"e-mail must be valid\\")
          );
        }"
            />
          </v-col>
        </v-row>
         
        <v-row
          class="ga__signup_icon_view__form_row"
          justify="center"
        >
          <v-col
            class="ga__signup_icon_view__form_col"
            cols="8"
          >
            <v-select
              color="teal"
              dark=""
              filled=""
              hint="Select the icons category in order to build up your pass"
              item-color="teal"
              items="buildings,astronomy,alphabet,animals,automative"
              label="icon category"
              persistent-hint=""
              required=""
              value="buildings"
            />
          </v-col>
        </v-row>
         
        <v-row
          class="hidden"
        >
          <v-col
            class="ga__signup_icon_view__icons_title"
            cols="12"
          >
            <div
              class="grey--text text--darken-1"
            >
              
                Choose icons for your
                
              <span
                class="text-overline font-weight-light white--text ga__signup_icon_view__icons_title__pass"
              >
                IPASS
              </span>
            </div>
             
            <!---->
             
            <!---->
          </v-col>
        </v-row>
         
        <v-row
          class="ga__signup_icon_view__icons_row hidden"
        >
          <v-col
            class="ga__signup_icon_view__icon"
            cols="2"
          >
            <div
              class="ga__signup_icon_view__icon__content"
            >
              <!---->
            </div>
          </v-col>
          <v-col
            class="ga__signup_icon_view__icon"
            cols="2"
          >
            <div
              class="ga__signup_icon_view__icon__content"
            >
              <!---->
            </div>
          </v-col>
          <v-col
            class="ga__signup_icon_view__icon"
            cols="2"
          >
            <div
              class="ga__signup_icon_view__icon__content"
            >
              <!---->
            </div>
          </v-col>
          <v-col
            class="ga__signup_icon_view__icon"
            cols="2"
          >
            <div
              class="ga__signup_icon_view__icon__content"
            >
              <!---->
            </div>
          </v-col>
          <v-col
            class="ga__signup_icon_view__icon"
            cols="2"
          >
            <div
              class="ga__signup_icon_view__icon__content"
            >
              <!---->
            </div>
          </v-col>
          <v-col
            class="ga__signup_icon_view__icon"
            cols="2"
          >
            <div
              class="ga__signup_icon_view__icon__content"
            >
              <!---->
            </div>
          </v-col>
        </v-row>
      </v-form>
    </v-col>
     
    <v-col
      class="ga__signup_icon_view__icons"
      cols="12"
      lg="6"
    >
      <v-row
        class="ga__signup_icon_view__icons__row"
      >
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__signup_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-archway"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__signup_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-building"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__signup_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-building-columns"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__signup_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-campground"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__signup_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-church"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__signup_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-city"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__signup_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-dungeon"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__signup_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-hospital"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__signup_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-house"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__signup_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-igloo"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__signup_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-shop"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__signup_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-vihara"
              size="2x"
            />
          </div>
        </v-col>
      </v-row>
    </v-col>
  </v-row>
   
  <div
    class="ga__signup_icon_view__logo_container"
  >
    <div
      class="ga__signup_icon_view__logo"
    />
  </div>
   
  <v-snackbar>
    
      Invalid username!
      
  </v-snackbar>
</div>
`;
